pd/src/testnet.rs

```diff
@@ -0,0 +1,671 @@
+use std::env::current_dir;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::{fmt, fs::File};
+
+use anyhow::Result;
+use directories::UserDirs;
+use regex::{Captures, Regex};
+use serde::{de, Deserialize};
+
+use penumbra_crypto::Address;
+use tendermint::PrivateKey;
+
+use crate::genesis;
+
+/// Methods and types used for generating testnet configurations.
+
+pub fn parse_allocations_file(input_file: PathBuf) -> Result<Vec<TestnetAllocation>> {
+    let file = match File::open(&input_file) {
+        Err(why) => panic!("couldn't open {}: {}", input_file.display(), why),
+        Ok(file) => file,
+    };
+
+    let mut rdr = csv::Reader::from_reader(file);
+    let mut res = vec![];
+    for result in rdr.deserialize() {
+        let record: TestnetAllocation = result?;
+        res.push(record);
+    }
+
+    Ok(res)
+}
+
+pub fn parse_validators_file(input_file: PathBuf) -> Result<Vec<TestnetValidator>> {
+    let file = match File::open(&input_file) {
+        Err(why) => panic!("couldn't open {}: {}", input_file.display(), why),
+        Ok(file) => file,
+    };
+
+    let validators: Vec<TestnetValidator> = serde_json::from_reader(file)?;
+
+    Ok(validators)
+}
+
+fn string_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    struct U64StringVisitor;
+
+    impl<'de> de::Visitor<'de> for U64StringVisitor {
+        type Value = u64;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("a string containing a u64 with optional underscores")
+        }
+
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            let r = v.clone().replace("_", "");
+            r.parse::<u64>().map_err(E::custom)
+        }
+
+        fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            Ok(v)
+        }
+    }
+
+    deserializer.deserialize_any(U64StringVisitor)
+}
+
+/// Hardcoded Tendermint config template. Should produce tendermint config similar to
+/// https://github.com/tendermint/tendermint/blob/6291d22f46f4c4f9121375af700dbdafa51577e7/cmd/tendermint/commands/init.go#L45
+/// There exists https://github.com/informalsystems/tendermint-rs/blob/a12118978f2ffea4042d6d38ebfb290d12611314/config/src/config.rs#L23 but
+/// this seemed more straightforward as only the moniker is changed right now.
+pub fn generate_tm_config(node_name: &str) -> String {
+    format!(
```

Comment:
This large format string can be put in a separate file by doing `format!(include_str!("tm_config_template.toml"), node_name)`.